### PR TITLE
Allow users to stash changes without switching branches

### PR DESCRIPTION
**Pull request by rafeca**
Closes https://github.com/desktop/desktop/issues/8197

Paired with @niik 

## Description

This PR adds the option to stash all local changes without having to checkout a different branch by adding two context menu items (one in the application menu, on the "Branch" submenu and the other on the context menu that appears when right-clicking on the changed files header:

Screen Shot 2020-06-17 at 18 51 43

Screen Shot 2020-06-17 at 18 51 07

This PR reuses the logic that keeps the changes on the current branch when checking out another PR, which means:

* The stash will get associated to the currently checked out branch (it will only be accessible from this branch within Desktop).
* If a stash already exists on the currently checked out branch, when stashing again it will get "overwritten" (a dialog will appear to inform the user about this).

### Product considerations

* The stash menu items will get disabled state when any of the following conditions is true:
  * There are no local changes.
  * The current HEAD is in a detached state (no branch currently checked out).
  * There is an unresolved git conflict.
* If there's an existing stash entry on the current branch, the following dialog will get shown (it's the same that gets shown when this scenario happens while switching branches):
* The <kbd>Cmd + Shift + S</kbd> keyboard shortcut has been added for this feature. It's not used anywhere in Desktop and by doing a quick search it doesn't seem to be used by global shortcuts in either macOS or Windows.

Screen Shot 2020-06-17 at 19 33 14

## Release notes

Notes: [Added] Allow users to stash changes without switching branches

### CODE REVIEW

**Comment by outofambit** [comment]
> remove the `...` from the references as there is no secondary action needed by the user, as it would imply.

that one's tricky, because the user _might_ get a popup asking if they want to replace the previous stash (if one already exists for that branch in desktop)

**Comment by maste197** [comment]
Ä``

File: app/src/ui/dispatcher/dispatcher.ts [hunk 1]
```diff
@@ -2325,6 +2325,24 @@ export class Dispatcher {
...
 
+  /**
+   * Creates a stash for the current branch. Note that this will
+   * override any stash that already exists for the current branch.
+   *
+   * @param repository
+   * @param showConfirmationDialog  Whether to show a confirmation dialog if an
+   *                                existing stash exists (defaults to true).
+   */
+  public createStashForCurrentBranch(
```
Comment (ayed42) [inline app/src/ui/dispatcher/dispatcher.ts:2336]: .742@![Uploading 90313977...]

Comment (ayed42) [reply app/src/ui/dispatcher/dispatcher.ts:2336]: we he w

**Pull request merged**